MEMMODEL_CONSTRAINTS.md

```diff
@@ -103,6 +103,19 @@ may perceive in that statement.)
   for memory used for atomics.  The situation is somewhat better in
   well-typed C++ but reality is complicated.
 
+## Implementation targeting C and C++ or LLVM
+
+* _Targeting C or C++_: it should be possible to implement
+  SharedArrayBuffer using C or C++.  The C/C++11 memory model allows
+  accesses with a number of orderings, and there should be a mapping
+  from SharedArrayBuffer accesses to C/C++11 orderings. Racy
+  SharedArrayBuffer accesses are required to produce defined values,
+  so cannot be implemented using C/C++ non-atomics.
+* _Targeting LLVM_: implementations may also decide to target
+  LLVM, which may provide for more optimization than targeting C/C++
+  and compiling to LLVM. In particular, LLVM allows unordered access,
+  which does not require per-variable sequential consistency.
```

Comment:
>     If we map SAB non-atomic accesses to LLVM's unordered, I think this all the same optimizations as LLVM non-atomic. Slightly annoyingly, C11's relaxed blocks some optimizations due to coherence.
> 
> unordered isn't the same as LLVM's non-atomics.

Indeed, I'm asking whether we can implement SAB non-atomic as LLVM unordered? (In particular, is there a requirement that SAB non-atomics are coherent?)

> relaxed basically only enforces same-address writes are seen in a thread's write order. You can totally reorder them, fuse them, move them around, eliminate dead ones, etc, and there's no guarantee that a relaxed write will ever be observed without synchronization actions occurring.

Relaxed operations can't be reordered as much as unordered operations, due to the requirement of coherence. For example, `x=1;x=2;x=3` can't be optimized to `x=2;x=1;x=3` for `relaxed` writes, but can for `unordered`.

>         Targeting C or C++: it should be possible to implement
>             SharedArrayBuffer using C or C++. The C/C++11 memory model allows
> 
> The wording is confusing: C/C++ should be mappable to SAB. Or are you saying that the implementation of SAB (e.g. a JS engine) is in C++? That's not really true either, because the SAB code is JIT-compiled.

I'm meaning the implementation of SAB (e.g. inside a JS engine). The question is, should it be possible to implement SAB in C++ or LLVM, or does the implementation have to be in hand-crafted asm? (This impacts my life, as someone who may be implementing the Servo interactions with SAB, if they end up being visible in the browser.)

>         accesses with a number of orderings, and there should be a mapping
>         from SharedArrayBuffer accesses to C/C++11 orderings. Racy
> 
> That's currently not true: only seq_cst is supported by SAB. There are issues open for other ones.

OK, what word should I use for non-atomic accesses to an SAB? Presumably non-atomic access is being mapped to something weaker than `seq_cst`.

>         SharedArrayBuffer accesses are required to produce defined values,
>         so cannot be implemented using C/C++ non-atomics.
> 
> That's not really accurate: they simply don't have the same semantics. C++ has much looser guarantees, both when it comes to races (they're simply UB) and alignment. As above I don't see SAB as being "implemented using C++".

Yes,. that's the point I'm trying to make here, SAB non-atomic accesses can't be mapped to C11 non-atomics, since racy non-atomics are UB in C11.

>     +* Targeting LLVM: implementations may also decide to target
> 
>         the LLVM memory model, for example by using a language such as
>         Rust, which inherits its memory model from LLVM. LLVM has a very
>         similar model to C/C++, but allows mixed-size access, and
>         allows Java-like unordered access.
> 
> I still don't think this is accurate. Implementations of what? Are we talking about Rust -> LLVM -> SAB, or about ??? -> SAB -> LLVM?

??? -> SAB -> LLVM.  This whole section is about how SAB can be implemented, not how to use it as a compilation target.

> I'd also like to clarify what you mean by Rust. Sure it uses LLVM, but I don't think it's accurate to say that its memory model is LLVM's.

OK, I talked to @nikomatsakis, and the Rust memory model is still a work in progress, so while the current implementation is via LLVM (and so inherits its memory model) future implementations may provide weaker guarantees.